File: python/packages/spore_tools/src/spore_tools/tools/git.py
"""Git tools: GitLog, GitDiff, GitCommit, GitStatus, GitReset."""

from __future__ import annotations

from spore_core.harness import (
    CommandOutput,
    SandboxProvider,
    ToolOutput,
    ToolOutputError,
    ToolOutputSuccess,
)
from spore_core.model import ToolCall
from spore_core.sandbox import SandboxViolationException
from spore_core.tool_registry import ToolAnnotations, ToolContext, ToolSchema

from ._common import finish_with_possible_truncation
from .error import SandboxViolationError, ToolExecutionError
from .params import (
    GitCommitParams,
    GitDiffParams,
    GitLogParams,
    GitResetMode,
    GitResetParams,
    GitStatusParams,
    parse_params,
)


async def _run_git(args: list[str], sandbox: SandboxProvider) -> CommandOutput:
    # SC-15: a git spawn failure surfaces from ``execute_command`` as a typed
    # ``SandboxViolationException`` (wrapping ``SandboxExecSpawnFailed``) rather
    # than a fake ``exit_code: -1``. Re-raise it as a ``SandboxViolationError``
    # (a ``ToolExecutionError``) so each tool's existing ``except
    # ToolExecutionError`` arm carries the typed violation to the harness, which
    # applies its ``SandboxViolationPolicy`` — recoverable feedback by default,
    # since ``exec_spawn_failed`` is Layer-2 (never halt-eligible). Mirrors
    # Rust's ``run_git`` → ``.map_err(ToolExecutionError::SandboxViolation)``.
    try:
        return await sandbox.execute_command("git", args, None, None)
    except SandboxViolationException as e:
        raise SandboxViolationError(violation=e.violation) from e


def _classify(out: CommandOutput) -> tuple[bool, str]:
    """Return ``(is_error, content_or_message)``."""

    if out.exit_code == 0:
        return False, out.stdout
    return True, f"git exit {out.exit_code} ; {out.stderr.rstrip()}"


class GitLogTool:
    NAME = "git_log"

    def name(self) -> str:
        return self.NAME

    def is_subagent_tool(self) -> bool:
        return False

    def may_produce_large_output(self) -> bool:
        return True

    @classmethod
    def schema(cls) -> ToolSchema:
        return ToolSchema(
            name=cls.NAME,
            description="Show recent git commits",
            parameters={
                "type": "object",
                "properties": {
                    "n": {"type": "integer"},
                    "format": {"type": "string"},
                },
            },
            annotations=ToolAnnotations(read_only=True),
        )

    async def execute(
        self, call: ToolCall, sandbox: SandboxProvider, ctx: ToolContext
    ) -> ToolOutput:
        try:
            params = parse_params(GitLogParams, call)
            args = ["log", "-n", str(params.n)]
            if params.format == "oneline":
                args.append("--oneline")
            else:
                args.append(f"--format={params.format}")
            out = await _run_git(args, sandbox)
        except ToolExecutionError as e:
            return e.to_tool_output()
        is_err, body = _classify(out)
        if is_err:
            return ToolOutputError(message=body, recoverable=True)
        return await finish_with_possible_truncation(body, call.id, sandbox)


class GitDiffTool:
    NAME = "git_diff"

    def name(self) -> str:
        return self.NAME

    def is_subagent_tool(self) -> bool:
        return False

    def may_produce_large_output(self) -> bool:
        return True

    @classmethod
    def schema(cls) -> ToolSchema:
        return ToolSchema(
            name=cls.NAME,
            description="Show a git diff",
            parameters={
                "type": "object",
                "properties": {
                    "from": {"type": "string"},
                    "to": {"type": "string"},
                },
            },
            annotations=ToolAnnotations(read_only=True),
        )

    async def execute(
        self, call: ToolCall, sandbox: SandboxProvider, ctx: ToolContext
    ) -> ToolOutput:
        try:
            params = parse_params(GitDiffParams, call)
            args = ["diff"]
            if params.from_ is not None:
                args.append(params.from_)
            if params.to is not None:
                args.append(params.to)
            out = await _run_git(args, sandbox)
        except ToolExecutionError as e:
            return e.to_tool_output()
        is_err, body = _classify(out)
        if is_err:
            return ToolOutputError(message=body, recoverable=True)
        return await finish_with_possible_truncation(body, call.id, sandbox)


class GitCommitTool:
    NAME = "git_commit"

    def name(self) -> str:
        return self.NAME

    def is_subagent_tool(self) -> bool:
        return False

    def may_produce_large_output(self) -> bool:
        return False

    @classmethod
    def schema(cls) -> ToolSchema:
        return ToolSchema(
            name=cls.NAME,
            description="Stage files (if any) and create a git commit",
            parameters={
                "type": "object",
                "properties": {
                    "message": {"type": "string"},
                    "files": {"type": "array", "items": {"type": "string"}},
                },
                "required": ["message"],
            },
            annotations=ToolAnnotations(destructive=True),
        )

    async def execute(
        self, call: ToolCall, sandbox: SandboxProvider, ctx: ToolContext
    ) -> ToolOutput:
        try:
            params = parse_params(GitCommitParams, call)
            combined = ""
            if params.files:
                add_out = await _run_git(["add", *params.files], sandbox)
                is_err, body = _classify(add_out)
                if is_err:
                    return ToolOutputError(message=body, recoverable=True)
                combined += body
            out = await _run_git(["commit", "-m", params.message], sandbox)
        except ToolExecutionError as e:
            return e.to_tool_output()
        is_err, body = _classify(out)
        if is_err:
            return ToolOutputError(message=body, recoverable=True)
        combined += body
        return ToolOutputSuccess(content=combined, truncated=False)


class GitStatusTool:
    NAME = "git_status"

    def name(self) -> str:
        return self.NAME

    def is_subagent_tool(self) -> bool:
        return False

    def may_produce_large_output(self) -> bool:
        return False

    @classmethod
    def schema(cls) -> ToolSchema:
        return ToolSchema(
            name=cls.NAME,
            description="Show git status (porcelain)",
            parameters={"type": "object", "properties": {}},
            annotations=ToolAnnotations(read_only=True),
        )

    async def execute(
        self, call: ToolCall, sandbox: SandboxProvider, ctx: ToolContext
    ) -> ToolOutput:
        # GitStatusParams accepts {} or extra-rejecting empty object.
        try:
            parse_params(GitStatusParams, call)
        except ToolExecutionError:
            # status takes no params — be lenient like the Rust reference.
            pass
        try:
            out = await _run_git(["status", "--porcelain"], sandbox)
        except ToolExecutionError as e:
            return e.to_tool_output()
        is_err, body = _classify(out)
        if is_err:
            return ToolOutputError(message=body, recoverable=True)
        return ToolOutputSuccess(content=body, truncated=False)


class GitResetTool:
    NAME = "git_reset"

    def name(self) -> str:
        return self.NAME

    def is_subagent_tool(self) -> bool:
        return False

    def may_produce_large_output(self) -> bool:
        return False

    @classmethod
    def schema(cls) -> ToolSchema:
        return ToolSchema(
            name=cls.NAME,
            description="Reset to a target commit (hard/soft/mixed)",
            parameters={
                "type": "object",
                "properties": {
                    "target": {"type": "string"},
                    "mode": {
                        "type": "string",
                        "enum": ["hard", "soft", "mixed"],
                    },
                },
                "required": ["target", "mode"],
            },
            annotations=ToolAnnotations(destructive=True),
        )

    async def execute(
        self, call: ToolCall, sandbox: SandboxProvider, ctx: ToolContext
    ) -> ToolOutput:
        try:
            params = parse_params(GitResetParams, call)
            flag = {
                GitResetMode.HARD: "--hard",
                GitResetMode.SOFT: "--soft",
                GitResetMode.MIXED: "--mixed",
            }[params.mode]
            out = await _run_git(["reset", flag, params.target], sandbox)
        except ToolExecutionError as e:
            return e.to_tool_output()
        is_err, body = _classify(out)
        if is_err:
            return ToolOutputError(message=body, recoverable=True)
        return ToolOutputSuccess(content=body, truncated=False)


__all__ = [
    "GitCommitTool",
    "GitDiffTool",
    "GitLogTool",
    "GitResetMode",
    "GitResetTool",
    "GitStatusTool",
]

File: go/spore-core/tools/git.go
// Git tools: GitLog, GitDiff, GitCommit, GitStatus, GitReset.
//
// Every git invocation goes through sandbox.ExecuteCommand — the tool never
// touches os/exec directly. That keeps the sandboxing contract intact.

package tools

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
)

// runGit shells out to `git <args...>` via the sandbox.
func runGit(
	ctx context.Context,
	args []string,
	sandbox sporecore.SandboxProvider,
) (sporecore.CommandOutput, *ToolExecutionError) {
	out, v := sandbox.ExecuteCommand(ctx, "git", args, "", 0)
	if v != nil {
		return sporecore.CommandOutput{}, SandboxViolationError(v)
	}
	return out, nil
}

// classify maps a git CommandOutput onto either its stdout (exit 0) or a
// recoverable error ToolOutput.
func classify(out sporecore.CommandOutput) (string, *sporecore.ToolOutput) {
	if out.ExitCode == 0 {
		return out.Stdout, nil
	}
	err := sporecore.ToolOutput{
		Kind:        sporecore.ToolOutputError,
		Message:     fmt.Sprintf("git exit %d ; %s", out.ExitCode, strings.TrimRight(out.Stderr, "\n")),
		Recoverable: true,
	}
	return "", &err
}

// ============================================================================
// GitLog
// ============================================================================

type GitLogTool struct{}

func NewGitLogTool() *GitLogTool { return &GitLogTool{} }

const GitLogToolName = "git_log"

func (*GitLogTool) Name() string                { return GitLogToolName }
func (*GitLogTool) IsSubagentTool() bool        { return false }
func (*GitLogTool) MayProduceLargeOutput() bool { return true }

func (*GitLogTool) Schema() sporecore.RegistryToolSchema {
	return sporecore.RegistryToolSchema{
		Name:        GitLogToolName,
		Description: "Show recent git commits",
		Parameters: json.RawMessage(`{
			"type": "object",
			"properties": {
				"n": {"type": "integer"},
				"format": {"type": "string"}
			}
		}`),
		Annotations: sporecore.ToolAnnotations{ReadOnly: true},
	}
}

func (t *GitLogTool) Execute(ctx context.Context, call sporecore.ToolCall, sandbox sporecore.SandboxProvider) sporecore.ToolOutput {
	params := GitLogParams{N: 20, Format: "oneline"}
	if len(call.Input) > 0 {
		if e := parseParams(call, &params); e != nil {
			return e.ToToolOutput()
		}
	}
	args := []string{"log", "-n", fmt.Sprintf("%d", params.N)}
	if params.Format == "oneline" {
		args = append(args, "--oneline")
	} else {
		args = append(args, "--format="+params.Format)
	}
	out, e := runGit(ctx, args, sandbox)
	if e != nil {
		return e.ToToolOutput()
	}
	body, errOut := classify(out)
	if errOut != nil {
		return *errOut
	}
	return finishWithPossibleTruncation(ctx, body, call.ID, sandbox)
}

// ============================================================================
// GitDiff
// ============================================================================

type GitDiffTool struct{}

func NewGitDiffTool() *GitDiffTool { return &GitDiffTool{} }

const GitDiffToolName = "git_diff"

func (*GitDiffTool) Name() string                { return GitDiffToolName }
func (*GitDiffTool) IsSubagentTool() bool        { return false }
func (*GitDiffTool) MayProduceLargeOutput() bool { return true }

func (*GitDiffTool) Schema() sporecore.RegistryToolSchema {
	return sporecore.RegistryToolSchema{
		Name:        GitDiffToolName,
		Description: "Show a git diff",
		Parameters: json.RawMessage(`{
			"type": "object",
			"properties": {
				"from": {"type": "string"},
				"to": {"type": "string"}
			}
		}`),
		Annotations: sporecore.ToolAnnotations{ReadOnly: true},
	}
}

func (t *GitDiffTool) Execute(ctx context.Context, call sporecore.ToolCall, sandbox sporecore.SandboxProvider) sporecore.ToolOutput {
	var params GitDiffParams
	if len(call.Input) > 0 {
		if e := parseParams(call, &params); e != nil {
			return e.ToToolOutput()
		}
	}
	args := []string{"diff"}
	if params.From != nil {
		args = append(args, *params.From)
	}
	if params.To != nil {
		args = append(args, *params.To)
	}
	out, e := runGit(ctx, args, sandbox)
	if e != nil {
		return e.ToToolOutput()
	}
	body, errOut := classify(out)
	if errOut != nil {
		return *errOut
	}
	return finishWithPossibleTruncation(ctx, body, call.ID, sandbox)
}

// ============================================================================
// GitCommit
// ============================================================================

type GitCommitTool struct{}

func NewGitCommitTool() *GitCommitTool { return &GitCommitTool{} }

const GitCommitToolName = "git_commit"

func (*GitCommitTool) Name() string                { return GitCommitToolName }
func (*GitCommitTool) IsSubagentTool() bool        { return false }
func (*GitCommitTool) MayProduceLargeOutput() bool { return false }

func (*GitCommitTool) Schema() sporecore.RegistryToolSchema {
	return sporecore.RegistryToolSchema{
		Name:        GitCommitToolName,
		Description: "Stage files (if any) and create a git commit",
		Parameters: json.RawMessage(`{
			"type": "object",
			"properties": {
				"message": {"type": "string"},
				"files": {"type": "array", "items": {"type": "string"}}
			},
			"required": ["message"]
		}`),
		Annotations: sporecore.ToolAnnotations{Destructive: true},
	}
}

func (t *GitCommitTool) Execute(ctx context.Context, call sporecore.ToolCall, sandbox sporecore.SandboxProvider) sporecore.ToolOutput {
	var params GitCommitParams
	if e := parseParams(call, &params); e != nil {
		return e.ToToolOutput()
	}
	combined := ""
	if len(params.Files) > 0 {
		args := append([]string{"add"}, params.Files...)
		out, e := runGit(ctx, args, sandbox)
		if e != nil {
			return e.ToToolOutput()
		}
		body, errOut := classify(out)
		if errOut != nil {
			return *errOut
		}
		combined += body
	}
	out, e := runGit(ctx, []string{"commit", "-m", params.Message}, sandbox)
	if e != nil {
		return e.ToToolOutput()
	}
	body, errOut := classify(out)
	if errOut != nil {
		return *errOut
	}
	combined += body
	return sporecore.ToolOutput{Kind: sporecore.ToolOutputSuccess, Content: combined}
}

// ============================================================================
// GitStatus
// ============================================================================

type GitStatusTool struct{}

func NewGitStatusTool() *GitStatusTool { return &GitStatusTool{} }

const GitStatusToolName = "git_status"

func (*GitStatusTool) Name() string                { return GitStatusToolName }
func (*GitStatusTool) IsSubagentTool() bool        { return false }
func (*GitStatusTool) MayProduceLargeOutput() bool { return false }

func (*GitStatusTool) Schema() sporecore.RegistryToolSchema {
	return sporecore.RegistryToolSchema{
		Name:        GitStatusToolName,
		Description: "Show git status (porcelain)",
		Parameters:  json.RawMessage(`{"type": "object", "properties": {}}`),
		Annotations: sporecore.ToolAnnotations{ReadOnly: true},
	}
}

func (t *GitStatusTool) Execute(ctx context.Context, call sporecore.ToolCall, sandbox sporecore.SandboxProvider) sporecore.ToolOutput {
	out, e := runGit(ctx, []string{"status", "--porcelain"}, sandbox)
	if e != nil {
		return e.ToToolOutput()
	}
	body, errOut := classify(out)
	if errOut != nil {
		return *errOut
	}
	return sporecore.ToolOutput{Kind: sporecore.ToolOutputSuccess, Content: body}
}

// ============================================================================
// GitReset
// ============================================================================

type GitResetTool struct{}

func NewGitResetTool() *GitResetTool { return &GitResetTool{} }

const GitResetToolName = "git_reset"

func (*GitResetTool) Name() string                { return GitResetToolName }
func (*GitResetTool) IsSubagentTool() bool        { return false }
func (*GitResetTool) MayProduceLargeOutput() bool { return false }

func (*GitResetTool) Schema() sporecore.RegistryToolSchema {
	return sporecore.RegistryToolSchema{
		Name:        GitResetToolName,
		Description: "Reset to a target commit (hard/soft/mixed)",
		Parameters: json.RawMessage(`{
			"type": "object",
			"properties": {
				"target": {"type": "string"},
				"mode": {"type": "string", "enum": ["hard", "soft", "mixed"]}
			},
			"required": ["target", "mode"]
		}`),
		Annotations: sporecore.ToolAnnotations{Destructive: true},
	}
}

func (t *GitResetTool) Execute(ctx context.Context, call sporecore.ToolCall, sandbox sporecore.SandboxProvider) sporecore.ToolOutput {
	var params GitResetParams
	if e := parseParams(call, &params); e != nil {
		return e.ToToolOutput()
	}
	var flag string
	switch params.Mode {
	case GitResetHard:
		flag = "--hard"
	case GitResetSoft:
		flag = "--soft"
	case GitResetMixed:
		flag = "--mixed"
	default:
		return InvalidParameters(fmt.Sprintf("unknown mode %q", params.Mode)).ToToolOutput()
	}
	out, e := runGit(ctx, []string{"reset", flag, params.Target}, sandbox)
	if e != nil {
		return e.ToToolOutput()
	}
	body, errOut := classify(out)
	if errOut != nil {
		return *errOut
	}
	return sporecore.ToolOutput{Kind: sporecore.ToolOutputSuccess, Content: body}
}

var (
	_ sporecore.Tool = (*GitLogTool)(nil)
	_ sporecore.Tool = (*GitDiffTool)(nil)
	_ sporecore.Tool = (*GitCommitTool)(nil)
	_ sporecore.Tool = (*GitStatusTool)(nil)
	_ sporecore.Tool = (*GitResetTool)(nil)
)
